execute/exectypes/type_optimizer.go

```diff
@@ -0,0 +1,209 @@
+	emptyCommitData := CommitData{}
+	emptySeqNr := make(map[cciptypes.SeqNum]cciptypes.Message)
+	emptySeqNrSize := 0
+
+	enc, err := json.Marshal(emptySeqNr)
+	if err == nil {
+		emptySeqNrSize = len(enc)
+	}
+
+	return EmptyEncodeSizes{
+		MessageAndTokenData: emptyMsg.EncodedSize() + emptyTokenData.EncodedSize(), // 397 + 18 = 415
+		CommitData:          emptyCommitData.EncodedSize(),                         // 305
+		SeqNumMap:           emptySeqNrSize,                                        // 2
+	}
+}
+
+// TruncateObservation truncates the observation to fit within the given op.maxEncodedSize after encoding.
+// It removes data from the observation in the following order:
+// For each chain, pick last report and start removing messages one at a time.
+// If removed all messages from the report, remove the report.
+// If removed last report in the chain, remove the chain.
+// After removing full report from a chain, move to the next chain and repeat. This ensures that we don't
+// exclude messages from one chain only.
+// Keep repeating this process until the encoded observation fits within the op.maxEncodedSize
+// Important Note: We can't delete messages completely from single reports as we need them to create merkle proofs.
+func (op ObservationOptimizer) TruncateObservation(observation Observation) (Observation, error) {
+	obs := observation
+	encodedObs, err := obs.Encode()
+	if err != nil {
+		return Observation{}, err
+	}
+	encodedObsSize := len(encodedObs)
+	if encodedObsSize <= op.maxEncodedSize {
+		return obs, nil
+	}
+
+	chains := maps.Keys(obs.CommitReports)
+	sort.Slice(chains, func(i, j int) bool {
+		return chains[i] < chains[j]
+	})
+
+	messageAndTokenDataEncodedSizes := GetEncodedMsgAndTokenDataSizes(obs.Messages, obs.TokenData)
+	// If the encoded obs is too large, start filtering data.
+	for encodedObsSize > op.maxEncodedSize {
+		for _, chain := range chains {
+			commits := obs.CommitReports[chain]
+			if len(commits) == 0 {
+				continue
+			}
+			lastCommit := &commits[len(commits)-1]
+			seqNum := lastCommit.SequenceNumberRange.Start()
+
+			for seqNum <= lastCommit.SequenceNumberRange.End() {
+				if _, ok := obs.Messages[chain][seqNum]; !ok {
+					return Observation{}, fmt.Errorf("missing message with seqNr %d from chain %d", seqNum, chain)
+				}
+				obs.Messages[chain][seqNum] = cciptypes.Message{}
+				obs.TokenData[chain][seqNum] = NewMessageTokenData()
+				// Subtract the removed message and token size
+				encodedObsSize -= messageAndTokenDataEncodedSizes[chain][seqNum]
+				// Add empty message and token encoded size
+				encodedObsSize += op.emptyEncodedSizes.MessageAndTokenData
+				seqNum++
+				if encodedObsSize <= op.maxEncodedSize {
+					return obs, nil
+				}
+			}
+
+			var sizeDelta int
+			// Reaching here means that all messages in the report are truncated, truncate the last commit
+			obs, sizeDelta = op.truncateLastCommit(obs, chain)
+
+			encodedObsSize += sizeDelta
```

Comment:
nit: rather than returning `delta`, maybe return `bytesTruncated`. That way you can subtract rather than adding a negative.